File: src/10-codegolf-expert/the-fall-episode-3-codesize/solution.py
import sys
input=sys.stdin.readline
RM={0:{},1:{'TOP':'BOTTOM','LEFT':'BOTTOM','RIGHT':'BOTTOM'},2:{'TOP':'RIGHT','RIGHT':'BOTTOM'},3:{'TOP':'LEFT','LEFT':'BOTTOM'},4:{'TOP':'LEFT','LEFT':'BOTTOM'},5:{'TOP':'RIGHT','RIGHT':'BOTTOM'},6:{'TOP':'RIGHT','LEFT':'BOTTOM'},7:{'TOP':'LEFT','RIGHT':'BOTTOM'},8:{'LEFT':'BOTTOM','RIGHT':'TOP'},9:{'RIGHT':'BOTTOM','LEFT':'TOP'},10:{'LEFT':'RIGHT','RIGHT':'LEFT'},11:{'TOP':'BOTTOM','BOTTOM':'TOP'},12:{'TOP':'RIGHT','RIGHT':'TOP','LEFT':'BOTTOM','BOTTOM':'LEFT'},13:{'TOP':'LEFT','LEFT':'TOP','RIGHT':'BOTTOM','BOTTOM':'RIGHT'}}
DX={'TOP':0,'BOTTOM':0,'LEFT':-1,'RIGHT':1};DY={'TOP':-1,'BOTTOM':1,'LEFT':0,'RIGHT':0}
OPP={'TOP':'BOTTOM','BOTTOM':'TOP','LEFT':'RIGHT','RIGHT':'LEFT'}
CW={'TOP':'RIGHT','RIGHT':'BOTTOM','BOTTOM':'LEFT','LEFT':'TOP'}
CCW={'TOP':'LEFT','LEFT':'BOTTOM','BOTTOM':'RIGHT','RIGHT':'TOP'}
def rot_room(rm,cw=True):
 R=CW if cw else CCW;nr={}
 for e,x in rm.items():nr[R[e]]=R[x]
 return nr
W,H=map(int,input().split())
T=[];LK=[]
for y in range(H):
 row=list(map(int,input().split()))
 T.append([abs(v)for v in row]);LK.append([v<0 for v in row])
EX=int(input())
rooms=[[dict(RM.get(T[y][x],{}))for x in range(W)]for y in range(H)]
rots={}
def gt(x,y):return rots.get((x,y),rooms[y][x])
def gx(rm,e):return rm.get(e)
def get_rotations(orig):
 r=[dict(orig)]
 for _ in range(3):r.append(rot_room(r[-1]));
 return r
def dfs(x,y,e):
 if y>=H:return[]
 if x<0 or x>=W or y<0:return None
 orig=rooms[y][x]
 if LK[y][x]:cands=[gt(x,y)]
 else:cands=get_rotations(orig)
 for c in cands:
  ex=gx(c,e)
  if not ex:continue
  nx,ny=x+DX[ex],y+DY[ex];ne=OPP[ex]
  prev=rots.get((x,y));rots[(x,y)]=c
  sub=dfs(nx,ny,ne)
  if sub is not None:
   if c!=rooms[y][x]:sub.append((x,y,c))
   return sub
  if prev is not None:rots[(x,y)]=prev
  else:rots.pop((x,y),None)
 return None
Q=[]
while 1:
 p=input().split();xi,yi=int(p[0]),int(p[1]);ent=p[2]
 R=int(input())
 for _ in range(R):input()
 if not Q:
  rots.clear();plan=dfs(xi,yi,ent)
  if plan:
   for x,y,tgt in plan:
    orig=rooms[y][x]
    r=0
    for i in range(4):
     test=dict(orig)
     for _ in range(i):test=rot_room(test)
     if test==tgt:r=i;break
    if r==0:continue
    d='RIGHT'if r<=2 else'LEFT';cnt=r if r<=2 else 4-r
    for _ in range(cnt):Q.append(f"{x} {y} {d}")
    rooms[y][x]=tgt
 if Q:print(Q.pop(0),flush=True)
 else:print("WAIT",flush=True)

File: src/10-codegolf-expert/the-fall-episode-3-codesize/solution.cpp
#include<cstdio>
#include<vector>
#include<string>
#include<map>
#include<cstring>
#include<set>
using namespace std;
int W,H,EX,T[22][22],LK[22][22];
map<int,int>RM[14];
int RG[14][4],RL[14];
int DX[]={0,1,0,-1},DY[]={-1,0,1,0},OP[]={2,3,0,1};
map<int,int>rot;
struct R{int x,y,t;};
vector<R>plan;
bool dfs(int x,int y,int e){
if(y>=H)return x==EX;
if(x<0||x>=W||y<0)return 0;
int o=T[y][x],k=y*W+x;
int nc=LK[y][x]?1:RL[o];
for(int ci=0;ci<nc;ci++){
int c=LK[y][x]?(rot.count(k)?rot[k]:o):RG[o][ci];
if(!RM[c].count(e))continue;
int ex=RM[c][e],nx=x+DX[ex],ny=y+DY[ex],ne=OP[ex];
int had=rot.count(k),pv=had?rot[k]:0;rot[k]=c;
if(dfs(nx,ny,ne)){if(c!=o)plan.push_back({x,y,c});return 1;}
if(had)rot[k]=pv;else rot.erase(k);}return 0;}
int main(){
RM[1][0]=2;RM[1][1]=2;RM[1][3]=2;RM[2][1]=3;RM[2][3]=1;RM[3][0]=2;RM[4][0]=3;RM[4][1]=2;RM[5][0]=1;RM[5][3]=2;RM[6][1]=3;RM[6][3]=1;RM[7][0]=2;RM[7][1]=2;RM[8][3]=2;RM[8][1]=2;RM[9][0]=2;RM[9][3]=2;RM[10][0]=3;RM[11][0]=1;RM[12][1]=2;RM[13][3]=2;
RL[0]=1;RG[0][0]=0;RL[1]=1;RG[1][0]=1;
RL[2]=2;RG[2][0]=2;RG[2][1]=3;RL[3]=2;RG[3][0]=3;RG[3][1]=2;
RL[4]=2;RG[4][0]=4;RG[4][1]=5;RL[5]=2;RG[5][0]=5;RG[5][1]=4;
for(int i=0;i<4;i++){RL[6+i]=4;int g[]={6,7,8,9};for(int j=0;j<4;j++)RG[6+i][j]=g[(j+i)%4];}
for(int i=0;i<4;i++){RL[10+i]=4;int g[]={10,11,12,13};for(int j=0;j<4;j++)RG[10+i][j]=g[(j+i)%4];}
scanf("%d%d",&W,&H);
for(int y=0;y<H;y++)for(int x=0;x<W;x++){int v;scanf("%d",&v);T[y][x]=v<0?-v:v;LK[y][x]=v<0;}
scanf("%d",&EX);
vector<string>Q;int EM[256];EM['T']=0;EM['R']=1;EM['L']=3;
for(;;){int xi,yi;char pos[9];scanf("%d%d%s",&xi,&yi,pos);int ent=EM[(int)pos[0]];
int R;scanf("%d",&R);for(int i=0;i<R;i++){int a,b;char p[9];scanf("%d%d%s",&a,&b,p);}
if(Q.empty()){rot.clear();plan.clear();
if(dfs(xi,yi,ent))for(auto&s:plan){int o=T[s.y][s.x],r=-1;
for(int i=0;i<RL[o];i++)if(RG[o][i]==s.t){r=i;break;}
if(r<=0)continue;char buf[30];
if(r<=2)for(int i=0;i<r;i++){sprintf(buf,"%d %d RIGHT",s.x,s.y);Q.push_back(buf);}
else{sprintf(buf,"%d %d LEFT",s.x,s.y);Q.push_back(buf);}
T[s.y][s.x]=s.t;}}
if(Q.size()){puts(Q[0].c_str());Q.erase(Q.begin());}else puts("WAIT");fflush(stdout);}}
